README.md
# egg bull board

<p align="center">
  <a href="https://github.com/vcapretz/bull-board/blob/master/LICENSE">
    <img alt="licence" src="https://img.shields.io/github/license/littleboyfury/egg-bull-board">
  </a>
  <img alt="open issues" src="https://img.shields.io/github/issues/littleboyfury/egg-bull-board"/>
  <a href="https://www.typescriptlang.org/">
    <img src="https://img.shields.io/badge/Typescript-4.3.5-green.svg" alt="typescript">
  </a>
  <a href="https://nodejs.org/en/">
    <img src="https://img.shields.io/badge/node->_v12-green.svg" alt="node">
  </a>
  <a href="https://www.ecma-international.org/publications-and-standards/standards/ecma-262/">
    <img src="https://img.shields.io/badge/Ecmascript-2019+-green.svg" alt="Ecmascript">
  </a>
<p>

## Install

```bash
npm install @skyfury/egg-bull-board -S
```

## bull-board

[bull-board](https://github.com/felixmosh/bull-board)

## Use

[with-egg](exmaple/with-egg/app/router.js)

```javascript
'use strict';

const Bull = require('bull')
const { EggAdapter } = require('@skyfury/egg-bull-board')
const { createBullBoard } = require('@bull-board/api');
const { BullAdapter } = require('@bull-board/api/bullAdapter');

const redisOptions = {
  port: 6379,
  host: 'localhost',
  password: '',
  tls: false,
};

const createQueue = (name) => new Bull(name, { connection: redisOptions });

/**
 * @param {Egg.Application} app - egg application
 */
module.exports = app => {
  const { router, controller } = app;

  const bullQueue = createQueue('bull');
  
  const serverAdapter = new EggAdapter();
  createBullBoard({
    queues: [new BullAdapter(bullQueue)],
    serverAdapter,
  });
  serverAdapter.setBasePath('/ui');
  app.use(serverAdapter.registerPlugin());
  
  router.get('/add', async (ctx) => {
    const opts = ctx.query.opts || {};

    if (opts.delay) {
      opts.delay = +opts.delay * 1000; // delay must be a number
    }

    await bullQueue.add('Add', { title: ctx.query.title }, opts);

    ctx.body = {
      ok: true,
    };
  });

  console.log(`For the UI of instance1, open http://localhost:${app.config.cluster.listen.port}/ui`);
  console.log('Make sure Redis is running on port 6379 by default');
  console.log('To populate the queue, run:');
  console.log(`  curl http://localhost:${app.config.cluster.listen.port}/add?title=Example`);
  console.log('To populate the queue with custom options (opts), run:');
  console.log(`  curl http://localhost:${app.config.cluster.listen.port}/add?title=Test&opts[delay]=9`);
};
```


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v4.0.1](https://github.com/felixmosh/bull-board/compare/v4.0.0...v4.0.1)

- Optional chain queue find [`b68cd36`](https://github.com/felixmosh/bull-board/commit/b68cd360d2cb34a75d2aba511d51ccb10c60f124)

#### [v4.0.0](https://github.com/felixmosh/bull-board/compare/v3.11.1...v4.0.0)

- chore: upgrade to fastify 4 [`1007e5d`](https://github.com/felixmosh/bull-board/commit/1007e5da6257b2eede7d824feca295f44019aa45)
- Create codeql-analysis.yml [`97e2fc8`](https://github.com/felixmosh/bull-board/commit/97e2fc8980b7618562e25ddca012909bb65b311b)
- Update README.md [`701e3d7`](https://github.com/felixmosh/bull-board/commit/701e3d79dee23096c6a9d342d03d19810cd3a3e8)

#### [v3.11.1](https://github.com/felixmosh/bull-board/compare/v3.11.0...v3.11.1)

- chore(deps): bump ejs from 3.1.6 to 3.1.7 in /examples/with-hapi [`bb87e77`](https://github.com/felixmosh/bull-board/commit/bb87e77e7778ceb4451bed32276bef15035b7c3a)
- chore(deps): bump ejs from 3.1.6 to 3.1.7 in /examples/with-hapi-auth [`74c246e`](https://github.com/felixmosh/bull-board/commit/74c246e15c37e51f7c38c336154a67aae8e2bd15)
- chore(deps): bump ejs from 3.1.6 to 3.1.7 in /examples/with-koa [`68350e5`](https://github.com/felixmosh/bull-board/commit/68350e54599f89cab1f8aefdffeef4744756f081)

#### [v3.11.0](https://github.com/felixmosh/bull-board/compare/v3.10.7...v3.11.0)

- Upgrade to @fastify/static, closes #408 [`#408`](https://github.com/felixmosh/bull-board/issues/408)
- Update fastify examples [`7770850`](https://github.com/felixmosh/bull-board/commit/7770850c241a32ef7cf53da177049a7a6f223306)
- Add connection example [`0d10d20`](https://github.com/felixmosh/bull-board/commit/0d10d20d5d3a071b22bacf8b293af2f239413ce7)
- Add package list to README.md [`c565593`](https://github.com/felixmosh/bull-board/commit/c565593d0681eae882cb4a41a30f78d4f0404140)

#### [v3.10.7](https://github.com/felixmosh/bull-board/compare/v3.10.6...v3.10.7)

- Change min support to 14, 16 in ci [`e3c2f6a`](https://github.com/felixmosh/bull-board/commit/e3c2f6a65c28383060ddc14c562f615a33a07678)

#### [v3.10.6](https://github.com/felixmosh/bull-board/compare/v3.10.5...v3.10.6)

> 24 April 2022

- Update lock file [`6ee3494`](https://github.com/felixmosh/bull-board/commit/6ee3494a8e2281ff66d8fec73d39e87a1e03d4d2)
- Bump deps [`0218db2`](https://github.com/felixmosh/bull-board/commit/0218db256f9b09234cbf2c6ebec3fc653a0e0e40)
- fix: resolve @types/react to v17 [`025b13e`](https://github.com/felixmosh/bull-board/commit/025b13e45cfc51bb705e3dee5a2f994d43ef33a9)

#### [v3.10.5](https://github.com/felixmosh/bull-board/compare/v3.10.4...v3.10.5)

> 24 April 2022

- Upgrade deps of examples [`27708de`](https://github.com/felixmosh/bull-board/commit/27708de288bf3f7cefea29df561c30fadd57afcb)
- Release 3.10.5 [`ce469b6`](https://github.com/felixmosh/bull-board/commit/ce469b657461c31a9da9e2b3a1e7ca2000a5d5e1)
- remove empty class [`1147472`](https://github.com/felixmosh/bull-board/commit/11474724c8f175968fcbd1e54f574b2ad9451be5)

#### [v3.10.4](https://github.com/felixmosh/bull-board/compare/v3.10.3...v3.10.4)

- Deps bump [`cefe20f`](https://github.com/felixmosh/bull-board/commit/cefe20f98dd24f0d0113fe43504455d453f28ef9)
- chore(deps): bump minimist from 1.2.5 to 1.2.6 [`7310de9`](https://github.com/felixmosh/bull-board/commit/7310de9c9cf417826d608a5a42b417e391626925)
- chore(deps): bump minimist in /examples/with-fastify-auth [`5c566d4`](https://github.com/felixmosh/bull-board/commit/5c566d406c431925be01a9dea46f1baa9795b419)

#### [v3.10.3](https://github.com/felixmosh/bull-board/compare/v3.10.2...v3.10.3)

> 8 April 2022

- fix: Remove the fade-out effect due to scrollbar issue, fixes #399 [`#399`](https://github.com/felixmosh/bull-board/issues/399)
- Release 3.10.3 [`9e34634`](https://github.com/felixmosh/bull-board/commit/9e346343fb4750e36f7a28ddebebb197c59e8bef)

#### [v3.10.2](https://github.com/felixmosh/bull-board/compare/v3.10.1...v3.10.2)

- fix: check if the job exists, closes #391 [`#391`](https://github.com/felixmosh/bull-board/issues/391)
- chore(deps): bump ansi-regex in /examples/with-fastify-auth [`f97564d`](https://github.com/felixmosh/bull-board/commit/f97564d76946d6cadb756cdfe70643b7fa8dae3f)
- chore(deps): bump node-forge from 1.2.1 to 1.3.0 [`f77de9b`](https://github.com/felixmosh/bull-board/commit/f77de9bb21b63968ee292b5d4c6ae4e9b07f04c2)

#### [v3.10.1](https://github.com/felixmosh/bull-board/compare/v3.10.0...v3.10.1)

- fix: align attemptsMade to the behavior of BullMQ. [`#386`](https://github.com/felixmosh/bull-board/issues/386)
- Bump deps [`bd7a6c0`](https://github.com/felixmosh/bull-board/commit/bd7a6c0923d41467a162c1c46a7e9dfbdd6856ce)
- Update ci steps to use node 14 + 16 [`5a6bd06`](https://github.com/felixmosh/bull-board/commit/5a6bd0639948044e24b1ab939ee53309fc361c05)
- Rollback pretty-bytes to v5 [`188fd98`](https://github.com/felixmosh/bull-board/commit/188fd982b103ba67ef29808c626d8f65eff03415)

#### [v3.10.0](https://github.com/felixmosh/bull-board/compare/v3.9.4...v3.10.0)

- Add canRetry options to queue [`#384`](https://github.com/felixmosh/bull-board/pull/384)
- chore(deps): bump follow-redirects from 1.14.7 to 1.14.8 [`#381`](https://github.com/felixmosh/bull-board/pull/381)
- Refactor tests to remove inline snapshots [`66ec345`](https://github.com/felixmosh/bull-board/commit/66ec34578ac0ba7c0483bc33ef84975733083d89)

#### [v3.9.4](https://github.com/felixmosh/bull-board/compare/v3.9.3...v3.9.4)

- Bump deps [`d4fdf02`](https://github.com/felixmosh/bull-board/commit/d4fdf02cd2326b1d6d3363ee1ce270c76b9db80f)
- fix test [`df80deb`](https://github.com/felixmosh/bull-board/commit/df80deb24b7637db38b2b15e04da4de619b00810)

src/EggAdapter.ts
import {
  AppControllerRoute,
  AppViewRoute,
  BullBoardQueues,
  ControllerHandlerReturnType,
  IServerAdapter,
} from '@bull-board/api/dist/typings/app'


import Koa from 'koa'
import Router from 'koa-router'
import mount from 'koa-mount'
import path from 'path'
import serve from 'koa-static'
import views from 'koa-views'

export class EggAdapter implements IServerAdapter {
  private basePath = ''
  private bullBoardQueues: BullBoardQueues | undefined
  private errorHandler: ((error: Error) => ControllerHandlerReturnType) | undefined
  private statics: { path: string; route: string } | undefined
  private viewPath: string | undefined
  private entryRoute: AppViewRoute | undefined
  private apiRoutes: AppControllerRoute[] | undefined

  // @bull-board/ui path
  private staticPath = path.resolve('node_modules/@bull-board/ui/dist')

  /**
   * base url
   * @param path
   */
  public setBasePath(path: string): EggAdapter {
    this.basePath = path
    return this
  }

  public setStaticPath(staticsRoute: string, staticsPath: string): EggAdapter {
    this.statics = { route: staticsRoute, path: staticsPath }

    return this
  }

  public setViewsPath(viewPath: string): EggAdapter {
    this.viewPath = viewPath
    return this
  }

  public setErrorHandler(handler: (error: Error) => ControllerHandlerReturnType) {
    this.errorHandler = handler
    return this
  }

  public setApiRoutes(routes: AppControllerRoute[]): EggAdapter {
    this.apiRoutes = routes

    return this
  }

  public setEntryRoute(routeDef: AppViewRoute): EggAdapter {
    this.entryRoute = routeDef

    return this
  }

  public setQueues(bullBoardQueues: BullBoardQueues): EggAdapter {
    this.bullBoardQueues = bullBoardQueues
    return this
  }

  public registerPlugin() {
    if (!this.statics) {
      throw new Error(`Please call 'setStaticPath' before using 'registerPlugin'`)
    } else if (!this.entryRoute) {
      throw new Error(`Please call 'setEntryRoute' before using 'registerPlugin'`)
    } else if (!this.viewPath) {
      throw new Error(`Please call 'setViewsPath' before using 'registerPlugin'`)
    } else if (!this.apiRoutes) {
      throw new Error(`Please call 'setApiRoutes' before using 'registerPlugin'`)
    } else if (!this.bullBoardQueues) {
      throw new Error(`Please call 'setQueues' before using 'registerPlugin'`)
    } else if (!this.errorHandler) {
      throw new Error(`Please call 'setErrorHandler' before using 'registerPlugin'`)
    }

    const app = new Koa()
    const router = new Router({
      strict: true,
    })

    app.use(async (ctx, next) => {
      try {
        await next()
      } catch (err) {
        if (this.errorHandler) {
          const { status, body } = this.errorHandler(err as Error)

          ctx.status = status || 500
          ctx.body = body
          ctx.app.emit('error', err, ctx)
        }
      }
    })

    app.use(
      views(this.viewPath, {
        extension: path.extname(this.entryRoute.handler().name).substring(1),
      }),
    )

    const { method, route, handler } = this.entryRoute
    const viewRoutes = Array.isArray(route) ? route : [route]
    viewRoutes.forEach((path) => {
      router[method](path, async (ctx) => {
        const { name } = handler();

        // rewrite render
        (ctx as any).render = views(this.staticPath, {
          extension: 'ejs',
          map: 'ejs'
        })(null as any, null as any)
        await (ctx as any).render(name, { basePath: this.basePath })
      })
    })

    app.use(mount(this.statics.route, serve(this.statics.path)))

    this.apiRoutes.forEach((route) => {
      const methods = Array.isArray(route.method) ? route.method : [route.method]
      methods.forEach((method) => {
        router[method](route.route, async (ctx) => {
          const response = await route.handler({
            queues: this.bullBoardQueues as any,
            params: ctx.params,
            query: ctx.query,
          })

          ctx.status = response.status || 200
          return (ctx.body = response.body)
        })
      })
    })

    app.use(router.routes()).use(router.allowedMethods())

    return mount(this.basePath || '', app)
  }
}


exmaple/with-egg/app/router.js
'use strict';

const Bull = require('bull')
const { EggAdapter } = require('@skyfury/egg-bull-board')
const { createBullBoard } = require('@bull-board/api');
const { BullAdapter } = require('@bull-board/api/bullAdapter');

const redisOptions = {
  port: 6379,
  host: 'localhost',
  password: '',
  tls: false,
};

const createQueue = (name) => new Bull(name, { connection: redisOptions });

/**
 * @param {Egg.Application} app - egg application
 */
module.exports = app => {
  const { router } = app;

  const bullQueue = createQueue('bull');
  const serverAdapter = new EggAdapter();

  createBullBoard({
    queues: [new BullAdapter(bullQueue)],
    serverAdapter,
  });

  serverAdapter.setBasePath('/ui');
  app.use(serverAdapter.registerPlugin());
  router.get('/add', async (ctx) => {
    const opts = ctx.query.opts || {};

    if (opts.delay) {
      opts.delay = +opts.delay * 1000; // delay must be a number
    }

    await bullQueue.add('Add', { title: ctx.query.title }, opts);

    ctx.body = {
      ok: true,
    };
  });

  console.log(`For the UI of instance1, open http://localhost:7001/ui`);
  console.log('Make sure Redis is running on port 6379 by default');
  console.log('To populate the queue, run:');
  console.log(`  curl http://localhost:7001/add?title=Example`);
  console.log('To populate the queue with custom options (opts), run:');
  console.log(`  curl http://localhost:7001/add?title=Test&opts[delay]=9`);
};
